src/lib/convert-forms.js

```diff
@@ -29,31 +28,35 @@ module.exports = (projectDir, subDirectory, options) => {
     return Promise.resolve();
   }
 
-  return fs.readdir(formsDir)
+  const candidateFiles = fs.readdir(formsDir)
     .filter(name => name.endsWith('.xlsx'))
     .filter(name => !name.startsWith('~$')) // ignore Excel "owner files"
-    .filter(name => name !== 'PLACE_TYPE-create.xlsx' && name !== 'PLACE_TYPE-edit.xlsx')
-    .filter(name => !options.forms || options.forms.includes(fs.withoutExtension(name)))
-    .reduce((promiseChain, xls) => {
-        const originalSourcePath = `${formsDir}/${xls}`;
-        let sourcePath;
-
-        if(options.force_data_node) {
-          const temporaryPath = `${fs.mkdtemp()}/${options.force_data_node}.xlsx`;
-          fs.copy(originalSourcePath, temporaryPath);
-          sourcePath = temporaryPath;
-        } else sourcePath = originalSourcePath;
-
-        const targetPath = `${fs.withoutExtension(originalSourcePath)}.xml`;
-
-        return promiseChain
-          .then(() => info('Converting form', originalSourcePath, '…'))
-          .then(() => xls2xform(sourcePath, targetPath))
-          .then(() => getHiddenFields(`${fs.withoutExtension(originalSourcePath)}.properties.json`))
-          .then(hiddenFields => fixXml(targetPath, hiddenFields, options.transformer, options.enketo))
-          .then(() => trace('Converted form', originalSourcePath));
-      },
-      Promise.resolve());
+    .filter(name => name !== 'PLACE_TYPE-create.xlsx' && name !== 'PLACE_TYPE-edit.xlsx');
+  
+  const formWhitelist = options && options.forms && options.forms.filter(form => !form.startsWith('--'));
+  const filteredFiles = candidateFiles.filter(name => !formWhitelist || !formWhitelist.length || formWhitelist.includes(fs.withoutExtension(name)));
```

Comment:
It seems a bit weird to iterate through the files if there is no whitelist. How about...

```js
const filteredFiles = (formWhitelist && formWhitelist.length) ? formWhitelist.includes(fs.withoutExtension(name)) : candidateFiles;
```

I think if it were me I would pull lines 31 to 41 out into a new function so it could do all the filtering and just return the result. The nice thing about that is you can return early if there's no whitelist so you don't need the ugly ternary.

Reply:
Good suggestion. Thanks